File: Examples/NoChat-Example/NoChat-Example/MMChatInputTextPanel.m
//
//  MMChatInputTextPanel.m
//  NoChat-Example
//
//  Copyright (c) 2016-present, little2s.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

#import "MMChatInputTextPanel.h"
#import "UIFont+NoChat.h"
#import <HPGrowingTextView/HPGrowingTextView.h>

#define MMRetinaPixel 0.5

@interface MMChatInputTextPanel () <HPGrowingTextViewDelegate>

@property (nonatomic, assign) UIEdgeInsets inputFieldInsets;
@property (nonatomic, assign) UIEdgeInsets inputFieldInternalEdgeInsets;
@property (nonatomic, assign) CGFloat baseHeight;

@property (nonatomic, assign) CGSize parentSize;

@property (nonatomic, assign) CGSize messageAreaSize;
@property (nonatomic, assign) CGFloat keyboardHeight;

@end

@implementation MMChatInputTextPanel

#pragma mark - Overrides

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        _baseHeight = 50;
        _inputFieldInsets = UIEdgeInsetsMake(7.5, 40, 7.5, 80);
        _inputFieldInternalEdgeInsets = UIEdgeInsetsMake(0, 4, 0, 0);
        
        _backgroundView = [[UIToolbar alloc] init];
        [self addSubview:_backgroundView];
        
        _fieldBackground = [[UIView alloc] init];
        _fieldBackground.frame = CGRectMake(40, 7.5, frame.size.width - 120, 35);
        _fieldBackground.backgroundColor = [UIColor whiteColor];
        _fieldBackground.layer.borderColor = [UIColor lightGrayColor].CGColor;
        _fieldBackground.layer.borderWidth = 0.5;
        _fieldBackground.layer.cornerRadius = 5;
        _fieldBackground.layer.masksToBounds = YES;
        [self addSubview:_fieldBackground];
        
        CGRect inputFieldClippingFrame = _fieldBackground.frame;
        _inputFieldClippingContainer = [[UIView alloc] initWithFrame:inputFieldClippingFrame];
        _inputFieldClippingContainer.clipsToBounds = YES;
        [self addSubview:_inputFieldClippingContainer];
        
        UIEdgeInsets inputFieldInternalEdgeInsets = _inputFieldInternalEdgeInsets;
        
        _inputField = [[HPGrowingTextView alloc] initWithFrame:CGRectMake(inputFieldInternalEdgeInsets.left, inputFieldInternalEdgeInsets.top, inputFieldClippingFrame.size.width - inputFieldInternalEdgeInsets.left, inputFieldClippingFrame.size.height)];
        _inputField.animateHeightChange = NO;
        _inputField.animationDuration = 0;
        _inputField.font = [UIFont systemFontOfSize:16];
        _inputField.backgroundColor = [UIColor clearColor];
        _inputField.opaque = NO;
        _inputField.clipsToBounds = YES;
        _inputField.internalTextView.backgroundColor = [UIColor clearColor];
        _inputField.internalTextView.opaque = NO;
        _inputField.internalTextView.contentMode = UIViewContentModeLeft;
        _inputField.internalTextView.enablesReturnKeyAutomatically = YES;
        _inputField.internalTextView.returnKeyType = UIReturnKeySend;
        _inputField.maxNumberOfLines = [self maxNumberOfLinesForSize:_parentSize];
        _inputField.delegate = self;
        
        _inputField.internalTextView.scrollIndicatorInsets = UIEdgeInsetsMake(-inputFieldInternalEdgeInsets.top, 0, 5 - MMRetinaPixel, 0);
        
        [_inputFieldClippingContainer addSubview:_inputField];
        
        _micButton = [UIButton buttonWithType:UIButtonTypeSystem];
        _micButton.exclusiveTouch = YES;
        [_micButton setImage:[UIImage imageNamed:@"MMVoice"] forState:UIControlStateNormal];
        [_micButton setImage:[UIImage imageNamed:@"MMVoiceHL"] forState:UIControlStateHighlighted];
        [self addSubview:_micButton];
        
        _faceButton = [UIButton buttonWithType:UIButtonTypeSystem];
        _faceButton.exclusiveTouch = YES;
        [_faceButton setImage:[UIImage imageNamed:@"MMEmotion"] forState:UIControlStateNormal];
        [_faceButton setImage:[UIImage imageNamed:@"MMEmotionHL"] forState:UIControlStateHighlighted];
        [self addSubview:_faceButton];
        
        _attachButton = [UIButton buttonWithType:UIButtonTypeSystem];
        _attachButton.exclusiveTouch = YES;
        [_attachButton setImage:[UIImage imageNamed:@"MMAttach"] forState:UIControlStateNormal];
        [_attachButton setImage:[UIImage imageNamed:@"MMAttachHL"] forState:UIControlStateHighlighted];
        [self addSubview:_attachButton];
    }
    return self;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    CGRect bounds = self.bounds;
    
    CGFloat baseHeight = self.baseHeight;
    
    self.backgroundView.frame = bounds;
    
    UIEdgeInsets inputFieldInsets = self.inputFieldInsets;
    self.fieldBackground.frame = CGRectMake(inputFieldInsets.left, inputFieldInsets.top, bounds.size.width - inputFieldInsets.left - inputFieldInsets.right, bounds.size.height - inputFieldInsets.top - inputFieldInsets.bottom);
    
    CGRect inputFieldClippingFrame = self.fieldBackground.frame;
    self.inputFieldClippingContainer.frame = inputFieldClippingFrame;
    
    self.micButton.frame = CGRectMake(0, bounds.size.height - baseHeight, 40, baseHeight);
    
    self.faceButton.frame = CGRectMake(bounds.size.width - 80, bounds.size.height - baseHeight, 40, baseHeight);
    
    self.attachButton.frame = CGRectMake(bounds.size.width - 40, bounds.size.height - baseHeight, 40, baseHeight);
}

- (void)endInputting:(BOOL)animated
{
    if (self.inputField.internalTextView.isFirstResponder) {
        [self.inputField.internalTextView resignFirstResponder];
    }
}

- (void)adjustForSize:(CGSize)size keyboardHeight:(CGFloat)keyboardHeight duration:(NSTimeInterval)duration animationCurve:(int)animationCurve
{
    CGSize previousSize = self.parentSize;
    self.parentSize = size;
    
    if (ABS(size.width - previousSize.width) > FLT_EPSILON) {
        [self changeToSize:size keyboardHeight:keyboardHeight duration:0];
    }
    
    [self adjustForSize:size keyboardHeight:keyboardHeight duration:duration inputFieldHeight:self.inputField.frame.size.height animationCurve:animationCurve];
}

- (void)changeToSize:(CGSize)size keyboardHeight:(CGFloat)keyboardHeight duration:(NSTimeInterval)duration
{
    self.parentSize = size;
    
    CGSize messageAreaSize = size;
    self.messageAreaSize = messageAreaSize;
    self.keyboardHeight = keyboardHeight;
    
    UIView *inputFieldSnapshotView = nil;
    if (duration > DBL_EPSILON) {
        inputFieldSnapshotView = [self.inputField.internalTextView snapshotViewAfterScreenUpdates:NO];
        inputFieldSnapshotView.frame = CGRectOffset(self.inputField.frame, self.inputFieldClippingContainer.frame.origin.x, self.inputFieldClippingContainer.frame.origin.y);
        [self addSubview:inputFieldSnapshotView];
    }
    
    [UIView performWithoutAnimation:^{
        [self updateInputFieldLayout];
    }];
    
    CGFloat inputContainerHeight = [self heightForInputFieldHeight:self.inputField.frame.size.height];
    CGRect newInputContainerFrame = CGRectMake(0, messageAreaSize.height - keyboardHeight - inputContainerHeight, messageAreaSize.width, inputContainerHeight);
    
    if (duration > DBL_EPSILON) {
        if (inputFieldSnapshotView != nil) {
            self.inputField.alpha = 0;
        }
        
        [UIView animateWithDuration:duration animations:^{
            self.frame = newInputContainerFrame;
            [self layoutSubviews];
            
            if (inputFieldSnapshotView != nil) {
                self.inputField.alpha = 1;
                inputFieldSnapshotView.frame = CGRectOffset(self.inputField.frame, self.inputFieldClippingContainer.frame.origin.x, self.inputFieldClippingContainer.frame.origin.y);
                inputFieldSnapshotView.alpha = 0;
            }
        } completion:^(BOOL finished) {
            [inputFieldSnapshotView removeFromSuperview];
        }];
    } else {
        self.frame = newInputContainerFrame;
    }
}

#pragma mark - Public

- (void)clearInputField
{
    self.inputField.internalTextView.text = nil;
}

#pragma mark - HPGrowingTextViewDelegate

- (void)growingTextView:(HPGrowingTextView *)growingTextView willChangeHeight:(float)height
{
    CGFloat inputContainerHeight = [self heightForInputFieldHeight:height];
    CGRect newInputContainerFrame = CGRectMake(0, self.messageAreaSize.height - self.keyboardHeight - inputContainerHeight, self.messageAreaSize.width, inputContainerHeight);
    
    [UIView animateWithDuration:0.3 animations:^{
        self.frame = newInputContainerFrame;
        [self layoutSubviews];
    }];
    
    id<MMChatInputTextPanelDelegate> delegate = (id<MMChatInputTextPanelDelegate>)self.delegate;
    if ([delegate respondsToSelector:@selector(inputPanel:willChangeHeight:duration:animationCurve:)]) {
        [delegate inputPanel:self willChangeHeight:inputContainerHeight duration:0.3 animationCurve:0];
    }
}

- (void)growingTextViewDidBeginEditing:(HPGrowingTextView *)growingTextView
{
    id<MMChatInputTextPanelDelegate> delegate = (id<MMChatInputTextPanelDelegate>)self.delegate;
    if ([delegate respondsToSelector:@selector(didInputTextPanelStartInputting:)]) {
        [delegate didInputTextPanelStartInputting:self];
    }
}

- (BOOL)growingTextView:(HPGrowingTextView *)growingTextView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text
{
    if (growingTextView != self.inputField) {
        return YES;
    }
    
    if ([text isEqualToString:@"\n"]) {
        NSString *str = [self.inputField.internalTextView.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
        if (str.length > 0) {
            id<MMChatInputTextPanelDelegate> delegate = (id<MMChatInputTextPanelDelegate>)self.delegate;
            if ([delegate respondsToSelector:@selector(inputTextPanel:requestSendText:)]) {
                [delegate inputTextPanel:self requestSendText:str];
            }
            [self clearInputField];
        }
        return NO;
    }
    
    return YES;
}

#pragma mark - Private

- (void)adjustForSize:(CGSize)size keyboardHeight:(CGFloat)keyboardHeight duration:(NSTimeInterval)duration inputFieldHeight:(CGFloat)inputFieldHeight animationCurve:(int)animationCurve
{
    dispatch_block_t block = ^{
        CGSize messageAreaSize = size;
        
        self.messageAreaSize = messageAreaSize;
        self.keyboardHeight = keyboardHeight;
        
        CGFloat inputContainerHeight = [self heightForInputFieldHeight:inputFieldHeight];
        self.frame = CGRectMake(0, messageAreaSize.height - keyboardHeight - inputContainerHeight, messageAreaSize.width, inputContainerHeight);
        [self layoutSubviews];
    };
    
    if (duration > DBL_EPSILON) {
        [UIView animateWithDuration:duration delay:0 options:animationCurve << 16 animations:block completion:nil];
    } else {
        block();
    }
}

- (CGFloat)heightForInputFieldHeight:(CGFloat)inputFieldHeight
{
    UIEdgeInsets inputFieldInsets = [self inputFieldInsets];
    CGFloat height = MAX([self baseHeight], inputFieldHeight - 8 + inputFieldInsets.top + inputFieldInsets.bottom);
    return height;
}

- (void)updateInputFieldLayout
{
    NSRange range = self.inputField.internalTextView.selectedRange;
    
    self.inputField.delegate = nil;
    
    UIEdgeInsets inputFieldInsets = [self inputFieldInsets];
    UIEdgeInsets inputFieldInternalEdgeInsets = [self inputFieldInternalEdgeInsets];
    
    CGRect inputFieldClippingFrame = CGRectMake(inputFieldInsets.left, inputFieldInsets.top, self.parentSize.width - inputFieldInsets.left - inputFieldInsets.right, 0);
    
    CGRect inputFieldFrame = CGRectMake(inputFieldInternalEdgeInsets.left, inputFieldInternalEdgeInsets.top, inputFieldClippingFrame.size.width - inputFieldInternalEdgeInsets.left, 0);
    
    self.inputField.frame = inputFieldFrame;
    self.inputField.internalTextView.frame = CGRectMake(0, 0, inputFieldFrame.size.width, inputFieldFrame.size.height);
    
    [self.inputField setMaxNumberOfLines:[self maxNumberOfLinesForSize:_parentSize]];
    [self.inputField refreshHeight];
    
    self.inputField.internalTextView.selectedRange = range;
    
    self.inputField.delegate = self;
}

- (int)maxNumberOfLinesForSize:(CGSize)size
{
    if (size.height <= 320.0f - FLT_EPSILON) {
        return 3;
    } else if (size.height <= 480.0f - FLT_EPSILON) {
        return 5;
    } else {
        return 7;
    }
}

@end

File: Examples/NoChat-Swift-Example/NoChat-Swift-Example/MMChatInputTextPanel.swift
//
//  MMChatInputTextPanel.swift
//  NoChat-Swift-Example
//
//  Copyright (c) 2016-present, little2s.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

import NoChat
import HPGrowingTextView

protocol MMChatInputTextPanelDelegate: NOCChatInputPanelDelegate {
    func didInputTextPanelStartInputting(_ inputTextPanel: MMChatInputTextPanel)
    func inputTextPanel(_ inputTextPanel: MMChatInputTextPanel, requestSendText text: String)
}

private let MMRetinaPixel = CGFloat(0.5)
private let MM_EPSILON = CGFloat(0.0001)

class MMChatInputTextPanel: NOCChatInputPanel, HPGrowingTextViewDelegate {
    
    var backgroundView: UIToolbar!
    
    var inputField: HPGrowingTextView!
    var inputFiledClippingContainer: UIView!
    var fieldBackground: UIView!
    
    var micButton: UIButton!
    var faceButton: UIButton!
    var attachButton: UIButton!
    
    private let inputFiledInsets = UIEdgeInsets(top: 7.5, left: 40, bottom: 7.5, right: 80)
    private let inputFiledInternalEdgeInsets = UIEdgeInsets(top: -3 - MMRetinaPixel, left: 0, bottom: 0, right: 0)
    private let baseHeight = CGFloat(50)
    
    private var parentSize = CGSize.zero
    
    private var messageAreaSize = CGSize.zero
    private var keyboardHeight = CGFloat(0)
    
    override init(frame: CGRect) {
        backgroundView = UIToolbar()
        
        fieldBackground = UIView()
        fieldBackground.frame = CGRect(x: 40, y: 7.5, width: frame.width - 120, height: 35)
        fieldBackground.backgroundColor = UIColor.white
        fieldBackground.layer.borderColor = UIColor.lightGray.cgColor
        fieldBackground.layer.borderWidth = 0.5
        fieldBackground.layer.cornerRadius = 5
        fieldBackground.layer.masksToBounds = true
        
        let inputFiledClippingFrame = CGRect(x: fieldBackground.frame.origin.x + 4, y: fieldBackground.frame.origin.y + 4, width: fieldBackground.frame.width - 8, height: fieldBackground.frame.height - 8)
        inputFiledClippingContainer = UIView(frame: inputFiledClippingFrame)
        inputFiledClippingContainer.clipsToBounds = true
        
        inputField = HPGrowingTextView(frame: CGRect(x: inputFiledInternalEdgeInsets.left, y: inputFiledInternalEdgeInsets.top, width: inputFiledClippingFrame.width - inputFiledInternalEdgeInsets.left, height: inputFiledClippingFrame.height))
        inputField.animateHeightChange = false
        inputField.animationDuration = 0
        inputField.font = UIFont.systemFont(ofSize: 16)
        inputField.backgroundColor = UIColor.clear
        inputField.isOpaque = false
        inputField.clipsToBounds = true
        inputField.internalTextView.backgroundColor = UIColor.clear
        inputField.internalTextView.isOpaque = false
        inputField.internalTextView.contentMode = .left
        inputField.internalTextView.enablesReturnKeyAutomatically = true
        inputField.internalTextView.returnKeyType = .send
        inputField.internalTextView.scrollIndicatorInsets = UIEdgeInsets(top: -inputFiledInternalEdgeInsets.top, left: 0, bottom: 5 - MMRetinaPixel, right: 0)
    
        micButton = UIButton(type: .system)
        micButton.isExclusiveTouch = true
        micButton.setImage(UIImage(named: "MMVoice"), for: .normal)
        micButton.setImage(UIImage(named: "MMVoiceHL"), for: .highlighted)
        
        faceButton = UIButton(type: .system)
        faceButton.isExclusiveTouch = true
        faceButton.setImage(UIImage(named: "MMEmotion"), for: .normal)
        faceButton.setImage(UIImage(named: "MMEmotionHL"), for: .highlighted)

        
        attachButton = UIButton(type: .system)
        attachButton.isExclusiveTouch = true
        attachButton.setImage(UIImage(named: "MMAttach")!, for: .normal)
        attachButton.setImage(UIImage(named: "MMAttachHL"), for: .highlighted)
        
        super.init(frame: frame)
        
        addSubview(backgroundView)
        addSubview(fieldBackground)
        addSubview(inputFiledClippingContainer)
        
        inputField.maxNumberOfLines = maxNumberOfLines(forSize: parentSize)
        inputField.delegate = self
        inputFiledClippingContainer.addSubview(inputField)
        
        addSubview(micButton)
        addSubview(faceButton)
        addSubview(attachButton)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        backgroundView.frame = bounds
        
        fieldBackground.frame = CGRect(x: 40, y: 7.5, width: bounds.width - 120, height: bounds.height - 15)
        
        inputFiledClippingContainer.frame = CGRect(x: fieldBackground.frame.origin.x + 4, y: fieldBackground.frame.origin.y + 4, width: fieldBackground.frame.width - 8, height: fieldBackground.frame.height - 8)
        
        micButton.frame = CGRect(x: 0, y: bounds.size.height - baseHeight, width: 40, height: baseHeight)
        
        faceButton.frame = CGRect(x: bounds.size.width - 80, y: bounds.size.height - baseHeight, width: 40, height: baseHeight)
        
        attachButton.frame = CGRect(x: bounds.size.width - 40, y: bounds.height - baseHeight, width: 40, height: baseHeight)
        
    }
    
    override func endInputting(_ animated: Bool) {
        if inputField.internalTextView.isFirstResponder {
            inputField.internalTextView.resignFirstResponder()
        }
    }
    
    override func adjust(for size: CGSize, keyboardHeight: CGFloat, duration: TimeInterval, animationCurve: Int32) {
        let previousSize = parentSize
        parentSize = size
        
        if abs(size.width - previousSize.width) > MM_EPSILON {
            change(to: size, keyboardHeight: keyboardHeight, duration: 0)
        }
        
        adjust(for: size, keyboardHeight: keyboardHeight, inputFiledHeight: inputField.frame.height, duration: duration, animationCurve: animationCurve)
    }
    
    override func change(to size: CGSize, keyboardHeight: CGFloat, duration: TimeInterval) {
        parentSize = size
        
        let messageAreaSize = size
        
        self.messageAreaSize = messageAreaSize
        self.keyboardHeight = keyboardHeight
        
        var inputFieldSnapshotView: UIView?
        if duration > .ulpOfOne {
            inputFieldSnapshotView = inputField.internalTextView.snapshotView(afterScreenUpdates: false)
            if let v = inputFieldSnapshotView {
                v.frame = inputField.frame.offsetBy(dx: inputFiledClippingContainer.frame.origin.x, dy: inputFiledClippingContainer.frame.origin.y)
                addSubview(v)
            }
        }
        
        UIView.performWithoutAnimation {
            self.updateInputFiledLayout()
        }
        
        let inputContainerHeight = heightForInputFiledHeight(inputField.frame.size.height)
        let newInputContainerFrame = CGRect(x: 0, y: messageAreaSize.height - keyboardHeight - inputContainerHeight, width: messageAreaSize.width, height: inputContainerHeight)
        
        if duration > .ulpOfOne {
            if inputFieldSnapshotView != nil {
                inputField.alpha = 0
            }
            
            UIView.animate(withDuration: duration, animations: {
                self.frame = newInputContainerFrame
                self.layoutSubviews()
                
                if let v = inputFieldSnapshotView {
                    self.inputField.alpha = 1
                    v.frame = self.inputField.frame.offsetBy(dx: self.inputFiledClippingContainer.frame.origin.x, dy: self.inputFiledClippingContainer.frame.origin.y)
                    v.alpha = 0
                }
            }, completion: { (_) in
                inputFieldSnapshotView?.removeFromSuperview()
            })
        } else {
            frame = newInputContainerFrame
        }
    }
    
    func clearInputField() {
        inputField.internalTextView.text = nil
        inputField.refreshHeight()
    }
    
    func growingTextView(_ growingTextView: HPGrowingTextView!, willChangeHeight height: Float) {
        let inputContainerHeight = heightForInputFiledHeight(CGFloat(height))
        let newInputContainerFrame = CGRect(x: 0, y: messageAreaSize.height - keyboardHeight - inputContainerHeight, width: messageAreaSize.width, height: inputContainerHeight)
        
        UIView.animate(withDuration: 0.3) {
            self.frame = newInputContainerFrame
            self.layoutSubviews()
        }
        
        delegate?.inputPanel(self, willChangeHeight: inputContainerHeight, duration: 0.3, animationCurve: 0)
    }
    
    func growingTextViewDidBeginEditing(_ growingTextView: HPGrowingTextView!) {
        if let d = delegate as? MMChatInputTextPanelDelegate {
            d.didInputTextPanelStartInputting(self)
        }
    }
    
    func growingTextView(_ growingTextView: HPGrowingTextView!, shouldChangeTextIn range: NSRange, replacementText text: String!) -> Bool {
        if growingTextView != self.inputField {
            return true
        }
        
        if text == "\n" {
            guard let txt = inputField.internalTextView.text else {
                return true
            }
            
            let str = txt.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
            if str.characters.count > 0 {
                if let d = delegate as? MMChatInputTextPanelDelegate {
                    d.inputTextPanel(self, requestSendText: str)
                }
                clearInputField()
            }
            
            return false
        }
        
        return true
    }
    
    private func adjust(for size: CGSize, keyboardHeight: CGFloat, inputFiledHeight: CGFloat, duration: TimeInterval, animationCurve: Int32) {
        let block = {
            let messageAreaSize = size
            
            self.messageAreaSize = messageAreaSize
            self.keyboardHeight = keyboardHeight
            
            let inputContainerHeight = self.heightForInputFiledHeight(inputFiledHeight)
            self.frame = CGRect(x: 0, y: messageAreaSize.height - keyboardHeight - inputContainerHeight, width: self.messageAreaSize.width, height: inputContainerHeight)
            self.layoutSubviews()
        }
        
        if duration > .ulpOfOne {
            UIView .animate(withDuration: duration, delay: 0, options: UIViewAnimationOptions(rawValue: UInt(animationCurve << 16)), animations: block, completion: nil)
        } else {
            block()
        }
    }
    
    private func heightForInputFiledHeight(_ inputFiledHeight: CGFloat) -> CGFloat {
        return max(baseHeight, inputFiledHeight - 8 + inputFiledInsets.top + inputFiledInsets.bottom + 8)
    }
    
    private func updateInputFiledLayout() {
        let range = inputField.internalTextView.selectedRange
        
        inputField.delegate = nil
        
        let inputFiledInsets = self.inputFiledInsets
        let inputFiledInternalEdgeInsets = self.inputFiledInternalEdgeInsets
        
        let inputFiledClippingFrame = CGRect(x: inputFiledInsets.left + 4, y: inputFiledInsets.top + 4, width: parentSize.width - inputFiledInsets.left - inputFiledInsets.right - 8, height: 0)
        
        let inputFieldFrame = CGRect(x: inputFiledInternalEdgeInsets.left, y: inputFiledInternalEdgeInsets.top, width: inputFiledClippingFrame.width - inputFiledInternalEdgeInsets.left, height: 0)
        
        inputField.frame = inputFieldFrame
        inputField.internalTextView.frame = CGRect(x: 0, y: 0, width: inputFieldFrame.width, height: inputFieldFrame.height)
        
        inputField.maxNumberOfLines = maxNumberOfLines(forSize: parentSize)
        inputField.refreshHeight()
        
        inputField.internalTextView.selectedRange = range
        
        inputField.delegate = self
    }
    
    private func maxNumberOfLines(forSize size: CGSize) -> Int32 {
        if size.height <= 320 {
            return 3
        } else if (size.height <= 480) {
            return 5;
        } else {
            return 7
        }
    }
    
}

